### PR TITLE
WT-8791 Give all tiered storage python tests a S3 scenario

### PR DESCRIPTION
**Pull request by sulabhM**

### CODE REVIEW

**Comment by sulabhM** [comment]
@sueloverso , @ddanderson Please start reviewing this change. I am still trying to get the runtime for the tests on evergreen to an acceptable value, but otherwise this change is ready to go in.

File: dist/api_data.py [hunk 1]
```diff
@@ -217,6 +217,8 @@ def __ge__(self, other):
             the bucket indicating the location for this table'''),
         Config('bucket_prefix', '', r'''
             the unique bucket prefix for this table'''),
+        Config('bucket_region', '', r'''
+            bucket region string identifier, represents the bucket's location'''),
```
Comment (sueloverso) [inline dist/api_data.py:221]: The "bucket" string was intended to incorporate the region (and anything else) into it by the user/caller. Is there any way the callers can just do that? Or does the S3 extension have to extract the region out specifically?

File: src/tiered/tiered_config.c [hunk 2]
```diff
@@ -99,6 +99,7 @@ __wt_tiered_bucket_config(
     WT_ERR(__wt_config_gets(session, cfg, "tiered_storage.bucket", &bucket));
     if (bucket.len == 0)
         WT_ERR_MSG(session, EINVAL, "table tiered storage requires bucket to be set");
+    WT_ERR(__wt_config_gets(session, cfg, "tiered_storage.bucket_region", &bucket_region));
```
Comment (sueloverso) [inline src/tiered/tiered_config.c:102]: Alphabetize its processing after `bucket_prefix`. 
Also I suggest a comment saying that it isn't required.

File: src/tiered/tiered_config.c [hunk 3]
```diff
@@ -169,6 +174,8 @@ __wt_tiered_conn_config(WT_SESSION_IMPL *session, const char **cfg, bool reconfi
     if (conn->bstorage == NULL)
         return (0);
     __wt_verbose(session, WT_VERB_TIERED, "TIERED_CONFIG: bucket %s", conn->bstorage->bucket);
+    __wt_verbose(
+      session, WT_VERB_TIERED, "TIERED_CONFIG: region %s", conn->bstorage->bucket_region);
```
Comment (sueloverso) [inline src/tiered/tiered_config.c:178]: Alphabetize after prefix.

File: src/tiered/tiered_config.c [hunk 4]
```diff
@@ -187,6 +194,7 @@ __wt_tiered_conn_config(WT_SESSION_IMPL *session, const char **cfg, bool reconfi
 err:
     __wt_free(session, conn->bstorage->auth_token);
     __wt_free(session, conn->bstorage->bucket);
+    __wt_free(session, conn->bstorage->bucket_region);
```
Comment (sueloverso) [inline src/tiered/tiered_config.c:197]: Alphabetize after prefix

File: test/suite/test_tiered02.py [hunk 5]
```diff
@@ -27,39 +27,64 @@
...
+    uri = "table:test_tiered02"
 
     def conn_config(self):
-        if not os.path.exists(self.bucket):
+        if self.ss_name == 'local_store' and not os.path.exists(self.bucket):
             os.mkdir(self.bucket)
         return \
           'tiered_storage=(auth_token=%s,' % self.auth_token + \
           'bucket=%s,' % self.bucket + \
+          'bucket_region=%s,' % self.bucket_region + \
```
Comment (sueloverso) [inline test/suite/test_tiered02.py:61]: Alphabetize after prefix

File: test/suite/test_tiered02.py [hunk 6]
```diff
@@ -27,39 +27,64 @@
...
             extlist.skip_if_missing = True
-        extlist.extension('storage_sources', self.extension_name)
+        extlist.extension('storage_sources', self.ss_name + config)
 
     def progress(self, s):
         self.verbose(3, s)
         self.pr(s)
 
     def confirm_flush(self, increase=True):
+        # Todo: Need to do something different for S3, disable for the time being
```
Comment (sueloverso) [inline test/suite/test_tiered02.py:84]: Is `Todo:` the right label? If there is a ticket, maybe use `FIXME-WT-<number>` or `TODO:`. I don't know the right answer here.
Also end the sentence comment with a period.

File: test/suite/test_tiered03.py [hunk 7]
```diff
@@ -41,41 +43,62 @@ class test_tiered03(wttest.WiredTigerTestCase):
     # sharing would probably need to be reworked.
     uri = 'file:test_tiered03'
 
+    storage_sources = [
+        ('local', dict(ss_name = 'local_store',
+            auth_token = get_auth_token('local_store'),
+            bucket = get_bucket1_name('local_store'),
+            bucket_region = get_bucket1_region('local_store'),
+            bucket_prefix = "pfx_")),
+        # WT-8896 - S3 extension doesnt gets stuck at initializing if more than one simultaneous
```
Comment (sueloverso) [inline test/suite/test_tiered03.py:52]: I think this should be `FIXME-WT-8896` (there might even be a script in `dist` that looks for `FIXME` on a ticket). Also the sentence does not quite make sense. I suggest rewording the beginning of the sentence maybe like:
"The S3 extension gets stuck during initialization if more..."

File: test/suite/test_tiered04.py [hunk 8]
```diff
@@ -28,11 +28,34 @@
...
 StorageSource = wiredtiger.StorageSource  # easy access to constants
 
 # test_tiered04.py
 #    Basic tiered storage API test.
 class test_tiered04(wttest.WiredTigerTestCase):
+    storage_sources = [
+        ('local', dict(ss_name = 'local_store',
+            auth_token = get_auth_token('local_store'),
+            bucket = get_bucket1_name('local_store'),
+            region = get_bucket1_region('local_store'),
```
Comment (sueloverso) [inline test/suite/test_tiered04.py:43]: Alphabetize region after prefix (in many places here)

File: test/suite/test_tiered04.py [hunk 9]
```diff
@@ -112,9 +138,10 @@ def test_tiered(self):
         conf = \
           ',tiered_storage=(auth_token=%s,' % self.auth_token + \
           'bucket=%s,' % self.bucket1 + \
+          'bucket_region=%s,' % self.region1 + \
```
Comment (sueloverso) [inline test/suite/test_tiered04.py:141]: Alphabetize

File: test/suite/test_tiered05.py [hunk 10]
```diff
@@ -26,34 +26,56 @@
...
 
 # test_tiered05.py
 #    Basic tiered storage API test error for tiered manager and flush_tier.
 class test_tiered05(wttest.WiredTigerTestCase):
-    uri = "table:test_tiered05"
+    storage_sources = [
+        ('local', dict(ss_name = 'local_store',
+            auth_token = get_auth_token('local_store'),
+            bucket = get_bucket1_name('local_store'),
+            bucket_region = get_bucket1_region('local_store'),
```
Comment (sueloverso) [inline test/suite/test_tiered05.py:42]: Alphabetize

File: test/suite/test_tiered08.py [hunk 11]
```diff
@@ -40,6 +42,20 @@
 #   Run background checkpoints and flush_tier operations while inserting
 #   data into a table from another thread.
 class test_tiered08(wttest.WiredTigerTestCase):
+    storage_sources = [
+        ('local', dict(ss_name = 'local_store',
+            auth_token = get_auth_token('local_store'),
+            bucket = get_bucket1_name('local_store'),
+            bucket_region = get_bucket1_region('local_store'),
```
Comment (sueloverso) [inline test/suite/test_tiered08.py:49]: Alphabetize

File: test/suite/test_tiered08.py [hunk 12]
```diff
@@ -48,27 +64,30 @@ class test_tiered08(wttest.WiredTigerTestCase):
...
-
     def conn_config(self):
-        if not os.path.exists(self.bucket):
+        if self.ss_name == 'local_store' and not os.path.exists(self.bucket):
             os.mkdir(self.bucket)
         return \
           'statistics=(fast),' + \
           'tiered_storage=(auth_token=%s,' % self.auth_token + \
           'bucket=%s,' % self.bucket + \
+          'bucket_region=%s,' % self.bucket_region + \
```
Comment (sueloverso) [inline test/suite/test_tiered08.py:74]: Alphabetize

File: test/suite/test_tiered10.py [hunk 13]
```diff
@@ -26,50 +26,80 @@
...
 # prefixes to the same bucket directory but different local databases.
 class test_tiered10(wttest.WiredTigerTestCase):
+    storage_sources = [
+        ('local', dict(ss_name = 'local_store',
+            auth_token = get_auth_token('local_store'),
+            bucket = get_bucket1_name('local_store'),
+            bucket_region = get_bucket1_region('local_store'),
+            prefix1 = '1_',
+            prefix2 = '2_')),
+        # WT-8896 - S3 extension doesnt gets stuck at initializing if more than one simultaneous
```
Comment (sueloverso) [inline test/suite/test_tiered10.py:46]: Same comment as before. I think `FIXME` is needed and same suggested wording change.

**Review by sueloverso — COMMENTED** [review]
I have several comments. Also `region` should sort/alphabetize after `prefix`. I noted many locations but gave up identifying them all somewhere around `test_tiered09.py`. All the tests needs to be fixed for that even if I didn't note them.

I have one fairly major question at the start - it was originally intended that the `bucket` string contained any needed information. Is the `region` necessary as a separate input or can it just be attached to the bucket string?

File: test/suite/test_tiered10.py (hunk 13 above)
Comment (ddanderson) [inline test/suite/test_tiered10.py:46]: Sentence needs rewording.

**Review by ddanderson — APPROVED** [review]
Nice work, this is looking good.  I noted one question and a couple small nits.  It's clear that these tests could use some refactoring to share duplicate code, I'll file a ticket.

File: dist/api_data.py (hunk 1 above)
Comment (sulabhM) [reply dist/api_data.py:221]: Hmm, having region as part of bucket name will actually make things simpler for the tiered storage and let extensions decide how they want to use that input. That makes sense. It should not be hard to make that change.

File: test/suite/test_tiered02.py (hunk 6 above)
Comment (etienneptl) [reply test/suite/test_tiered02.py:84]: FIXME with the correct ticket number is the way to go :)

File: test/suite/test_tiered03.py (hunk 7 above)
Comment (etienneptl) [reply test/suite/test_tiered03.py:52]: FYI, the script is [s_mentions](https://github.com/wiredtiger/wiredtiger/blob/develop/dist/s_mentions).

File: test/suite/helper_tiered.py [hunk 14]
```diff
@@ -50,42 +50,26 @@ def get_auth_token(storage_source):
...
-def get_bucket_list(storage_source):
-    if storage_source == 'local_store':
-        return local_buckets
-    if storage_source == 's3_store':
-        return s3_buckets
-    return None
+local_buckets = ['bucket1', 'bucket2']
 
 # Get name of the first bucket in the list.
 def get_bucket1_name(storage_source):
```
Comment (etienneptl) [inline test/suite/helper_tiered.py:59]: Instead of defining `get_bucket1_name` and `get_bucket2_name`, can we define `get_bucket_name(storage_source, index)`?

Comment (etienneptl) [reply test/suite/helper_tiered.py:59]: As discussed, will be fixed in WT-8855.

**Review by etienneptl — APPROVED** [review]
Good job, LGTM!

File: test/suite/helper_tiered.py (hunk 14 above)
Comment (sulabhM) [reply test/suite/helper_tiered.py:59]: I had thought so initially, but I don't imagine using more than 2 buckets, and where the function gets called, it looks cleaner and more readable:
```
get_bucket_name('s3_store', 1)
```
vs
```
get_bucket1_name('s3_store')
```

It is hard to figure out what that 1 means without a comment, and it is hard to put too many comments in scenario decit code.

File: test/suite/helper_tiered.py [hunk 15]
```diff
@@ -65,25 +96,13 @@ def generate_s3_prefix(test_name = ''):
...
 
-    if test_name:
-        prefix += test_name + '--'
+    # If the calling function has not provided a name, extract it from the stack.
+    # It is important to generate unique prefixes for different tests in the same class,
+    # so that the database namespace do not collide.
+    # 0th element on the stack is the current function. 1st element is the calling function.
+    if not test_name:
+        test_name = inspect.stack()[1][3]
+    prefix += test_name + '--'
```
[the remark this answers is not in the document]
Comment (sulabhM) [reply test/suite/helper_tiered.py:89]: Line number 81 generates a 32-bit random number, that takes care of the same user running the tests multithreaded. A current time (second granularity) is also prefixed. So I doubt there will ever be a collision.

This part of generating prefixes need a cleanup. I will like to improve it a bit, so that all the tests could go to the S3 bucket in a directory like hierarchy. And artefacts from same tests could be bunched together. I also need to figure out a way to archive artefacts from failures. I have these items noted down, will file a ticket shortly.

File: test/suite/test_tiered03.py [hunk 16]
```diff
@@ -41,41 +42,59 @@ class test_tiered03(wttest.WiredTigerTestCase):
...
-    cache_dir = "mybucket-cache" # a relative pathname, it will not be shared
-    bucket_prefix = "pfx_"
-    extension_name = "local_store"
 
     def conn_config(self):
-        # We have multiple connections that want to share a bucket directory.
-        # The first time this function is called, we'll establish the absolute
-        # path for the bucket, and always use that for the bucket name.
+        bucket_ret = self.bucket
+        cache_dir = self.bucket + '-cache'
```
Comment (sulabhM) [inline test/suite/test_tiered03.py:66]: Need to strip region from the bucket before appending -cache

**Comment by ddanderson** [comment]
I've filed WT-8905 to explore refactoring opportunities after this is merged.

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 17]
```diff
@@ -486,30 +486,46 @@ S3CustomizeFileSystem(WT_STORAGE_SOURCE *storageSource, WT_SESSION *session, con
 
     s3 = (S3_STORAGE *)storageSource;
 
-    /* We need to have a bucket to setup the file system. */
-    if (bucketName == NULL || strlen(bucketName) == 0) {
+    /*
+     * We need to have a bucket to setup the file system. The bucket is expected to be a name and a
+     * region, separated by a semi-colon. eg: 'abcd;ap-southeast-2'.
```
Comment (sueloverso) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:491]: This seems fine to me. I have a few questions/comments about it:
1. I don't know anything about the conventions around S3 buckets and regions to know: Would it make more sense to have it region;bucket instead? It seems region is the broader piece and bucket is finer-grained. Anyone have opinions?
2. Is a semi-colon the typical delimiter? I originally mis-read that as a colon because that is what I seemed to expect. Again, I don't know conventions. Are there conventions? If so, we should follow them.
3. Is this expectation documented anywhere? It needs to be somewhere in the documentation to make sure callers can find this information.

File: test/suite/test_tiered03.py [hunk 18]
```diff
@@ -41,41 +42,65 @@ class test_tiered03(wttest.WiredTigerTestCase):
     # sharing would probably need to be reworked.
     uri = 'file:test_tiered03'
 
+    storage_sources = [
+        ('local', dict(auth_token = get_auth_token('local_store'),
+            bucket = get_bucket1_name('local_store'),
+            bucket_prefix = "pfx_",
+            ss_name = 'local_store')),
+        # FIXME-WT-8896 The S3 extension gets stuck during initialization if more than one
+        # simultaneous WT connection is created. Renable once we have fixed this issue.
```
Comment (sueloverso) [inline test/suite/test_tiered03.py:51]: Renable is not a word. It could be just "Enable". That would make it consistent with the other test with the similar comment.

File: test/suite/test_tiered10.py [hunk 19]
```diff
@@ -26,50 +26,76 @@
...
 # prefixes to the same bucket directory but different local databases.
 class test_tiered10(wttest.WiredTigerTestCase):
+    storage_sources = [
+        ('local', dict(auth_token = get_auth_token('local_store'),
+            bucket = get_bucket1_name('local_store'),
+            prefix1 = '1_',
+            prefix2 = '2_',
+            ss_name = 'local_store')),
+        # FIXME-WT-8896 The S3 extension gets stuck during initialization if more than one
+        # simultaneous WT connection is created. Renable once we have fixed this issue.
```
Comment (sueloverso) [inline test/suite/test_tiered10.py:45]: Same issue here with Renable.

**Review by sueloverso — APPROVED** [review]
LGTM - I have a couple comments/questions about conventions. But otherwise this is looking good.

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 17 above)
Comment (sulabhM) [reply ext/storage_sources/s3_store/s3_storage_source.cpp:491]: I am not aware of the conventions either. I tried to find them, but most documents are around a legal alphabet set for the bucket name. ';' will work for now. I will file a ticket to talk to other (cloud) teams that know better. Also, there is a documentation ticket WT-8135, and I will file another documentation ticket to address the information required to build and use S3 extension.

File: test/suite/test_tiered10.py (hunk 19 above)
Comment (sulabhM) [reply test/suite/test_tiered10.py]: fixed

File: test/suite/test_tiered03.py (hunk 18 above)
Comment (sulabhM) [reply test/suite/test_tiered03.py]: fixed

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 17 above)
Comment (sulabhM) [reply ext/storage_sources/s3_store/s3_storage_source.cpp:491]: I created WT-8907 to both look for standards around bucket naming. I will probably talk to cloud team for that. The ticket will also update documentation.
